### DISCUSSION TITLE
How can I select a specific service / port?

### DISCUSSION THREAD

**Discussion by cetetesoft** [phpro/soap-client#557] [Q&A]
Hello.

I am currently using a WSDL with two services, each consisting of four ports. After generating all the required code with `vendor/bin/soap-client wizard`, all the requests are using the first endpoint (1st service / 1st port).

I guess I could achieve this by overriding code such as `CodeGeneratorEngineFactory`, creating a `Wsdl1MetadataProvider` instance with a different `ServiceSelectionCriteria`. Is there an easier way to do this and prevent so much duplicate code?

I'm quite new to both this library and SOAP, so I'm not sure if all of this make sense.

Thanks in advance.

**Comment by veewee** [comment]
Hello,

This will become easier in v4, since it will be possible to specify it from the Factory options instead.
You can try out the pre-releases of v4 which will be stable soon.

See:

* https://github.com/phpro/soap-client/blob/v4.x/src/Phpro/SoapClient/Soap/EngineOptions.php
* https://github.com/phpro/soap-client/blob/v3.x/src/Phpro/SoapClient/Soap/CodeGeneratorEngineFactory.php

Currently, the only way to filter ports is by using a preferred soap version, but I suppose filtering by port name / address or something like that would be rather easy to implement as well:

See:

* https://github.com/php-soap/wsdl-reader/blob/80c3f578d728db1a0d26e8d1412b88f43883f417/src/Model/Definitions/Ports.php#L33-L50

**Comment by cetetesoft** [comment]
Thank you.

Do you mean it will be possible to do so in the stable release of v4? I don't see a simple way to pick a specific port under 4.0.0-alpha3. Filtering by SOAP version is not an option for me, since all ports are SOAP 1.1.

Adding a `$name` property to `ServiceSelectionCriteria` would indeed do the trick.

**Reply by veewee** [reply]
It's currently not implemented yet but could indeed be added.

Would that 'name' then filter on service name or rather on port name? Or should a filter option be added for both?
It might be easier to see what you want to do exactly by sharing your wsdl service section.

**Reply by cetetesoft** [reply]
A port filter would be enough for me. These are the services and ports as defined on the WSDL I'm working with (as you can see, placeholders are being used for locations which are not yet available):
```xml
<wsdl:service name="sfVerifactu">
  <wsdl:port name="SistemaVerifactu" binding="sfWdsl:sfVerifactu">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
  <wsdl:port name="SistemaVerifactuSello" binding="sfWdsl:sfVerifactu">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
  <wsdl:port name="SistemaVerifactuPruebas" binding="sfWdsl:sfVerifactu">
    <soap:address location="https://prewww1.aeat.es/wlpl/TIKE-CONT/ws/SistemaFacturacion/VerifactuSOAP"/>
  </wsdl:port>
  <wsdl:port name="SistemaVerifactuSelloPruebas" binding="sfWdsl:sfVerifactu">
    <soap:address location="https://prewww10.aeat.es/wlpl/TIKE-CONT/ws/SistemaFacturacion/VerifactuSOAP"/>
  </wsdl:port>
</wsdl:service>
<wsdl:service name="sfRequerimiento">
  <wsdl:port name="SistemaRequerimiento" binding="sfWdsl:sfRequerimiento">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
  <wsdl:port name="SistemaRequerimientoSello" binding="sfWdsl:sfRequerimiento">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
  <wsdl:port name="SistemaRequerimientoPruebas" binding="sfWdsl:sfRequerimiento">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
  <wsdl:port name="SistemaRequerimientoSelloPruebas" binding="sfWdsl:sfRequerimiento">
    <soap:address location="xxxxxxxxxx"/>
  </wsdl:port>
</wsdl:service>
```

I need to target the 3rd port from both services (`SistemaVerifactuPruebas` and `SistemaRequerimientoPruebas`) for testing. After deploying my code to production, I'll need the 1st port from both (`SistemaVerifactu` and `SistemaRequerimiento`). Because port names are all different across services, I don't need a service filter, although it might be useful for someone else.

**Reply by veewee** [reply]
Thanks for your answer. I think adding both a `serviceName` and `portName` to `ServiceSelectionCriteria` would be the way to go. 

I'll make sure this is possible in v4 soon.

**Reply by veewee** [reply]
@cetetesoft 
Can you validate this PR works for your case? https://github.com/phpro/soap-client/pull/558
If so, I'll tag another alpha release.

(Relies on https://github.com/php-soap/wsdl-reader/releases/tag/0.19.0)

**Reply by cetetesoft** [reply]
The `ServiceSelectionCriteria` class seems to be missing some methods: `withServiceName()` and `withPort()`. Did you forget to include it in the PR?

**Reply by veewee** [reply]
It's part of https://github.com/php-soap/wsdl-reader/releases/tag/0.19.0. You might need to bump that dependency.

**Reply by cetetesoft** [reply]
Ah, yes, you're right...
However, I'm now facing a new problem (with both the version from the PR and `4.0.0-alpha3`). When performing the same test I used for version 3, I'm getting an exception:
`Uncaught Soap\Engine\Exception\MetadataException: No SOAP type found with name any.`
I've repeated the test under a new project and running `vendor/bin/soap-client wizard` again. The exception persists. Do you have any idea of what's going on? Should I post this as a new issue?

**Reply by veewee** [reply]
This looks like a new issue indeed. 
Can you create one?

Please also explain:

* A copy of the WSDL would be very handy to reproduce 
* Where do you get this exception: is it during code generation or rather during request encoding or response decoding? A stack trace might be handy as well

**Reply by cetetesoft** [reply]
I guess it's request encoding. I've posted this new one as issue #559.

**Reply by cetetesoft** [reply]
After forcing version 0.4.6 on package `goetas/xsd-reader`, the code is working fine. It's selecting the right location. Thanks!

**Comment by rdiazAicor** [comment]
Hi,

I had the same problem as you.
To fix it, you can pass the location parameter to the array of options used in the factory:

![location_parameter_soap](https://github.com/user-attachments/assets/55573cb1-8c2a-430e-8118-cabd3e1c228b)

I hope this helps :)

**Reply by veewee** [reply]
@rdiazAicor Thanks for your answer. 

Note that this won't work anymore for next version 4, since this package will not be using PHP's soap extension anymore.
You'dd either have to select the exact port or overwrite the base uri with [an HTTP plugin](https://docs.php-http.org/en/latest/plugins/request-uri-manipulations.html).

It is also limited to one service this way. Given the example above, there are multiple services with multiple ports and bindings available. Only overriding the address location wouldn't be enough in that specific situation.

**Reply by rdiazAicor** [reply]
Thank you for your reply also veewee.
I will take into account what you say when using version 4 :)
